app/modules/asset.rb

```diff
@@ -0,0 +1,148 @@
+class Asset
+
+  DEFAULT_START_DATE = 8.days.ago.strftime('%F')
+  DEFAULT_END_DATE = 1.day.ago.strftime('%F')
+
+
+  attr_reader :basic_stats,
+              :sources,
+              :os_sources,
+              :traffic_sources,
+              :country_sources,
+              :service,
+              :params,
+              :property,
+              :current_start_date,
+              :current_end_date,
+              :previous_period,
+              :current_period
+
+
+  def initialize(service, params)
+    @service = service
+    @params = params
+    @property = get_property
+    @basic_stats = get_basic_stats
+    @sources = get_sources
+    @os_sources = get_os_sources
+    @traffic_sources = get_traffic_sources
+    @country_sources = get_country_sources
+    @previous_pariod, @current_period = basic_stats.rows.each_slice(basic_stats.rows.size/2).to_a
+    @current_start_date = set_date(current_period[0][0])
+    @current_end_date = set_date(current_period[-1][0])
```

Comment:
I see your point. I considered this approach but decided agains it. Good to get a second opinion and I concur.